Add multi-system to FEProblemSolve and deploy on Physics syntax

GiudGiud opened the pull request:
closes #29019 

Includes:
- support for setting system names in Physics, which can then be coded to assign the variables to the right systems. There is no matching variables to systems from the input directly. I dont think all options are needed for the main class.
- a reworked FEProblemSolve that includes the fixed point iteration (or once-through loop) using a CV object for each system
- fixes for residual debugging for multi-system
- support for using the dynamic pressure for solving the NS equations. This one is a small change but consequential and I need it for the same project as this. 

Next items we need (TODO but not in this PR?)
- Nonlinear convergence for each NL system that don't fight over the EquationSystems parameter (currently an issue)
This is requiring this PR in libmesh:  https://github.com/libMesh/libmesh/pull/4007
- A decent convergence object for multi-system fixed point. Right now if you tried to use the nonlinear residual (like for multiapp iterations), you'd end up with always considering it converged (because you just solved that nonlinear system, but did not recompute the residual after solving the other nonlinear systems)
- Relatedly, an execute_on for multisystem fixed-point convergence checks (needed to update any PP during that convergence check) or for post system1 solve (a la post-Executor style)

For these last two tasks, i think the solutions will be fairly close to what gets used for MultiAppFixedPointConvergence when that gets figured out.

GiudGiud committed:
Add support for multi-system at the Physics base class refs #8780 #29019

GiudGiud committed:
Enable multi-system in derived physics in MOOSE

GiudGiud committed:
Add testing for multi-system + physics

GiudGiud committed:
Add a new solve object for solves with multiple systems

GiudGiud committed:
Add support for multiple nonlinear cv objects for mnultisys for default cv behavior

GiudGiud committed:
Add multi-nonlinear system support to FEProblemSolve

GiudGiud committed:
Add support for multi-nonlinear systems residual debugging

GiudGiud committed:
Offer option to solve for dynamic pressure from the Physics

GiudGiud committed:
Evaluate fluid properties using the total pressure

GiudGiud committed:
Add test for using the dynamic pressure

GiudGiud committed:
Add a test for picard multi-system
refs #8780

GiudGiud committed:
Improve classic multi-system error message
Add documentation for multi-system fixed point

joshuahansel reviewed (changes requested):
It's looking good!

GiudGiud commented on framework/doc/content/source/executioners/FEProblemSolve.md:
```diff
@@ -3,7 +3,22 @@
+to let the `FEProblemSolve` know when to terminate the fixed point loop.
+
+!alert note
+Options are currently limited for setting a multi-system fixed point convergence. We do not recommend using the
+nonlinear residual with a [VariableResidual.md] postprocessor or a [DefaultNonlinearConvergence.md] as these
+are not re-computed the end of a multi-system fixed point iteration.
```
I think that's what people would try

unfortunately no good option right now

GiudGiud replied:
We should contribute something that works soon though. Once multiapp-fp is figured out I think it will be trivial how to do that one

GiudGiud commented on framework/src/physics/PhysicsBase.C:
```diff
@@ -244,6 +250,19 @@ PhysicsBase::initializePhysics()
+    if (std::find(problem_systems.begin(), problem_systems.end(), sys_name) ==
+            problem_systems.end() &&
+        solverVariableNames().size())
+      mooseError("System '", sys_name, "' is not used in the problem");
```
there's a linear physics for flow already yes. 

I ll see if we re missing something

GiudGiud replied:
added linear

GiudGiud commented on framework/include/physics/PhysicsBase.h:
```diff
@@ -158,8 +167,11 @@ class PhysicsBase : public Action, public InputParametersChecksUtils<PhysicsBase
+  std::vector<SolverSystemName> _system_names;
+
+  /// System numbers for the system(s) owning the solver variables
+  std::vector<unsigned int> _system_numbers;
```
the number is for all systems (so across all physics) whereas the index in this vector would only cover this Physics

GiudGiud commented on framework/src/executioners/FEProblemSolve.C:
```diff
@@ -254,60 +266,169 @@ FEProblemSolve::FEProblemSolve(Executioner & ex)
+
+  // Outer loop for multi-grid convergence
+  bool converged = false;
+  unsigned int num_fp_multisys = 0;
```
there are fp_iterations for multiapps so I d avoid dropping multisys. I ll at _iter

joshuahansel replied:
Is there ambiguity with multiapps in this context? If yes, `multisys_fp_iter`. If no, `fp_iter`.

joshuahansel replied:
The `num` part read to me as the number of systems for some reason.

GiudGiud replied:
with _iter as the suffix now it should be fine?

GiudGiud commented on framework/include/executioners/FEProblemSolve.h:
```diff
@@ -38,8 +38,14 @@ class FEProblemSolve : public NonlinearSolveObject
+  /// @param index index of the nonlinear system
+  /// @return parameter for that nonlinear system
+  template <typename T>
+  T getParamFromNonLinearSystemVectorParam(std::string param_name, unsigned int index) const;
```
I ll do the lower case. I like that it starts with getParam like the regular one, but then says why there is a vector

GiudGiud commented on framework/src/problems/FEProblemBase.C:
```diff
@@ -8764,11 +8773,20 @@ FEProblemBase::resizeMaterialData(const Moose::MaterialDataType data_type,
+std::vector<ConvergenceName>
+FEProblemBase::getNonlinearConvergenceNames() const
 {
   if (_set_nonlinear_convergence_name)
```
done

GiudGiud commented on modules/contact/src/problems/AugmentedLagrangianContactProblem.C:
```diff
@@ -67,7 +67,8 @@ AugmentedLagrangianContactProblemTempl<ReferenceResidualProblem>::addDefaultNonl
   params.applyParameters(parameters());
   params.set<bool>("added_as_default") = true;
-  this->addConvergence(class_name, this->getNonlinearConvergenceName(), params);
+  // TODO: Add multi-nonlinear system support
```
done

GiudGiud committed:
Adress Josh s review

- more careful distinction between linear and nonlinear systems
- fix error messages

Co-authored-by: joshuahansel <joshua.hansel@inl.gov>

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/2543836/), step Clang format on abbcc10 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/29021/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/29021/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 681f13dbddd9572f6024c14e8561ea3f10447ab2`

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/2544017/), step Docs: sync website on b2069b1 wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/29021/site)

This comment will be updated on new commits.

GiudGiud committed:
Make framework tests pass:
- remove extra space on Solve  converged
- check for the existence of a single default NL conv
- make the minimalApp test more resilient to ordering of object creation

GiudGiud committed:
Skip RC UO execution on addition NL systems

GiudGiud committed:
Fix multi-system restart from physics

GiudGiud committed:
Add additional API to facilitate getting the system number from the variable name
Re-order tasks to make adding variables before checking integrity in physics (checks variables & systems)
Add missing required tasks

GiudGiud committed:
Allow 1D full-dimensional elements in a 2D mesh for projections

GiudGiud committed:
Add legacy single-system behavior back to allow Optim modules executioners to function
They rely on solving only one system, then performing operations in their OTHER solveObject (not fixed point solve, not inner_solve)
then solving the second one

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/2552542/), step Generate coverage on 4c83d80 wanted to post the following:


## Framework coverage

<table>
  <thead>
    <tr>
      <th></th>
      <th>61ed70</th>
      <th colspan=4>#29021 4c83d8</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>85.13%</td>
      <td>85.10%</td>
      <td>-0.03%</td>
      <td>74.00%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>107285</td>
      <td>107393</td>
      <td>+108</td>
      <td>185</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>18743</td>
      <td>18800</td>
      <td>+57</td>
      <td>65</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/29021/coverage_diff/framework)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/29021/coverage/framework)


## Modules coverage

### Contact

<table>
  <thead>
    <tr>
      <th></th>
      <th>61ed70</th>
      <th colspan=4>#29021 4c83d8</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>90.41%</td>
      <td>90.38%</td>
      <td>-0.03%</td>
      <td>66.67%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>4957</td>
      <td>4959</td>
      <td>+2</td>
      <td>4</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>526</td>
      <td>528</td>
      <td>+2</td>
      <td>2</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/29021/coverage_diff/contact)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/29021/coverage/contact)

### Heat transfer

<table>
  <thead>
    <tr>
      <th></th>
      <th>61ed70</th>
      <th colspan=4>#29021 4c83d8</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>88.81%</td>
      <td>88.81%</td>
      <td>+0.00%</td>
      <td>100.00%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>4404</td>
      <td>4406</td>
      <td>+2</td>
      <td>2</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>555</td>
      <td>555</td>
      <td>-</td>
      <td>0</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/29021/coverage_diff/heat_transfer)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/29021/coverage/heat_transfer)

### Navier stokes

<table>
  <thead>
    <tr>
      <th></th>
      <th>61ed70</th>
      <th colspan=4>#29021 4c83d8</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>84.64%</td>
      <td>84.74%</td>
      <td>+0.10%</td>
      <td>96.71%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>17081</td>
      <td>17282</td>
      <td>+201</td>
      <td>353</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>3100</td>
      <td>3112</td>
      <td>+12</td>
      <td>12</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/29021/coverage_diff/navier_stokes)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/29021/coverage/navier_stokes)

### Optimization

<table>
  <thead>
    <tr>
      <th></th>
      <th>61ed70</th>
      <th colspan=4>#29021 4c83d8</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>88.54%</td>
      <td>88.55%</td>
      <td>+0.01%</td>
      <td>100.00%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>1971</td>
      <td>1973</td>
      <td>+2</td>
      <td>2</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>255</td>
      <td>255</td>
      <td>-</td>
      <td>0</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/29021/coverage_diff/optimization)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/29021/coverage/optimization)

### Scalar transport

<table>
  <thead>
    <tr>
      <th></th>
      <th>61ed70</th>
      <th colspan=4>#29021 4c83d8</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>88.86%</td>
      <td>88.89%</td>
      <td>+0.03%</td>
      <td>100.00%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>367</td>
      <td>368</td>
      <td>+1</td>
      <td>1</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>46</td>
      <td>46</td>
      <td>-</td>
      <td>0</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/29021/coverage_diff/scalar_transport)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/29021/coverage/scalar_transport)

## Full coverage reports

<details>
<summary>Reports</summary>

- [`framework`](https://mooseframework.inl.gov/docs/PRs/29021/coverage/framework)
- [`chemical_reactions`](https://mooseframework.inl.gov/docs/PRs/29021/coverage/chemical_reactions)
- [`combined`](https://mooseframework.inl.gov/docs/PRs/29021/coverage/combined)
- [`contact`](https://mooseframework.inl.gov/docs/PRs/29021/coverage/contact)
- [`electromagnetics`](https://mooseframework.inl.gov/docs/PRs/29021/coverage/electromagnetics)
- [`external_petsc_solver`](https://mooseframework.inl.gov/docs/PRs/29021/coverage/external_petsc_solver)
- [`fluid_properties`](https://mooseframework.inl.gov/docs/PRs/29021/coverage/fluid_properties)
- [`fsi`](https://mooseframework.inl.gov/docs/PRs/29021/coverage/fsi)
- [`functional_expansion_tools`](https://mooseframework.inl.gov/docs/PRs/29021/coverage/functional_expansion_tools)
- [`geochemistry`](https://mooseframework.inl.gov/docs/PRs/29021/coverage/geochemistry)
- [`heat_transfer`](https://mooseframework.inl.gov/docs/PRs/29021/coverage/heat_transfer)
- [`level_set`](https://mooseframework.inl.gov/docs/PRs/29021/coverage/level_set)
- [`misc`](https://mooseframework.inl.gov/docs/PRs/29021/coverage/misc)
- [`navier_stokes`](https://mooseframework.inl.gov/docs/PRs/29021/coverage/navier_stokes)
- [`optimization`](https://mooseframework.inl.gov/docs/PRs/29021/coverage/optimization)
- [`peridynamics`](https://mooseframework.inl.gov/docs/PRs/29021/coverage/peridynamics)
- [`phase_field`](https://mooseframework.inl.gov/docs/PRs/29021/coverage/phase_field)
- [`porous_flow`](https://mooseframework.inl.gov/docs/PRs/29021/coverage/porous_flow)
- [`ray_tracing`](https://mooseframework.inl.gov/docs/PRs/29021/coverage/ray_tracing)
- [`rdg`](https://mooseframework.inl.gov/docs/PRs/29021/coverage/rdg)
- [`reactor`](https://mooseframework.inl.gov/docs/PRs/29021/coverage/reactor)
- [`richards`](https://mooseframework.inl.gov/docs/PRs/29021/coverage/richards)
- [`scalar_transport`](https://mooseframework.inl.gov/docs/PRs/29021/coverage/scalar_transport)
- [`solid_mechanics`](https://mooseframework.inl.gov/docs/PRs/29021/coverage/solid_mechanics)
- [`solid_properties`](https://mooseframework.inl.gov/docs/PRs/29021/coverage/solid_properties)
- [`stochastic_tools`](https://mooseframework.inl.gov/docs/PRs/29021/coverage/stochastic_tools)
- [`thermal_hydraulics`](https://mooseframework.inl.gov/docs/PRs/29021/coverage/thermal_hydraulics)
- [`xfem`](https://mooseframework.inl.gov/docs/PRs/29021/coverage/xfem)

</details>

## Warnings

- `framework` new line coverage rate 74.00% is less than the suggested 90.0%
- `contact` new line coverage rate 66.67% is less than the suggested 90.0%



This comment will be updated on new commits.

GiudGiud commented:
Got to add doco and fix the odd straggler tests but otherwise ready to get looked at

GiudGiud committed:
Add parameter groups, improve logic when using legacy single-system behavior

GiudGiud committed:
Add another error to cover wrong size for vector params

GiudGiud committed:
Add more exception testing

GiudGiud commented on modules/optimization/src/executioners/SteadyAndAdjoint.C:
```diff
@@ -23,6 +23,13 @@ SteadyAndAdjoint::validParams()
+  // We seek to prevent the MultiSystemSolveObject from solving both systems
+  // This is abusing input parameters, but SolveObjects do not have their own syntax
+  // and we need to send this parameter from the executioner to the default nested SolveObject
+  params.addPrivateParam<bool>("solve_only_first_system", true);
```
could also be cleaner to set the system_names here and not make them providate. I have not made up my mind

lindsayad commented on framework/src/executioners/FEProblemSolve.C:
gonna be substantial merge conflicts for me 😢

lindsayad commented on framework/include/executioners/FEProblemSolve.h:
```diff
@@ -38,8 +40,14 @@ class FEProblemSolve : public NonlinearSolveObject
+  /// @param index index of the nonlinear system
+  /// @return parameter for that nonlinear system
+  template <typename T>
+  T getParamFromNonlinearSystemVectorParam(std::string param_name, unsigned int index) const;
```
```suggestion
  T getParamFromNonlinearSystemVectorParam(const std::string & param_name, unsigned int index) const;
```

lindsayad commented on framework/src/executioners/MultiSystemSolveObject.C:
```diff
@@ -0,0 +1,60 @@
+    {
+      if (std::find(nl_sys_names.begin(), nl_sys_names.end(), sys_name) != nl_sys_names.end())
+      {
+        _systems.push_back(&_problem.getSolverSystem(_problem.nlSysNum(sys_name)));
+        _num_nl_systems++;
+      }
+      else if (std::find(linear_sys_names.begin(), linear_sys_names.end(), sys_name) !=
+               linear_sys_names.end())
+        _systems.push_back(&_problem.getSolverSystem(_problem.nlSysNum(sys_name)));
```
why not store these separately and avoid the dynamic cast in `FEProblemSolve`?

lindsayad replied:
I think it's better becase you are making two different `FEProblemBase::solveXXX` calls depending on the system type so you are getting no benefit from having this abstraction

GiudGiud replied:
One benefit is that I am naturally keeping the order from the user input in how I m solving these.
When the user specifies vector parameters for a mixed system-type it will matter

lindsayad replied:
that's true. Ok

lindsayad commented on test/tests/minimal_app/tests:
```diff
@@ -6,7 +6,7 @@
     type = RunApp
     cli_args = '--minimal --list-constructed-objects'
-    expect_out = "FEProblem\nDefaultNonlinearConvergence\nGeneratedMesh"
+    expect_out = "FEProblem\n*GeneratedMesh"
```
You could just move the convergence object ahead of the problem

GiudGiud replied:
I saw someone on discussion get a test failure there. I think this test is brittle

lindsayad replied:
Yes it's because @joshuahansel must have done a `git sed` to change the name and then didn't change the order here

lindsayad commented on framework/include/executioners/FEProblemSolve.h:
```diff
@@ -51,4 +59,11 @@ class FEProblemSolve : public NonlinearSolveObject
+  const bool _using_multi_sys_fp_iterations;
+  /// Convergence object to assess the convergence of the multi-system fixed point iteration
+  Convergence * _multi_sys_fp_convergence;
+  /// Whether to only solve the first nonlinear system, which is the legacy behavior
+  const bool _solve_only_first_system;
```
I don't like this data member. If this can be true, then this class is not truly a `MultiSystemSolveObject`. I think either this data member needs to go away or `FEProblemSolve` should not inherit from `MultiSystemSolveObject`

GiudGiud replied:
Ok that's fair. I did not need this, until I tried to get the optimization module tests to pass. It uses two solve objects m, one for each system.

I have a different solution you might like more.

GiudGiud replied:
Removed

lindsayad commented on framework/include/executioners/FEProblemSolve.h:
```diff
@@ -51,4 +59,11 @@ class FEProblemSolve : public NonlinearSolveObject
+
+  /// Whether we are using fixed point iterations for multi-system
+  const bool _using_multi_sys_fp_iterations;
+  /// Convergence object to assess the convergence of the multi-system fixed point iteration
+  Convergence * _multi_sys_fp_convergence;
```
If we had a class dedicated to multisystem solves then this could be a reference. Trying to retro-fit `FEProblemSolve` as a multi-system solve really is feeling like a poor fit. 

On a different note, I do not know if there are segregated solve options that do not fit in a fixed-point class of algorithms? In which case there may not be a need for this `_fp_` denotion

lindsayad replied:
This also takes it down the path of objects like `IterationAdaptiveDT` where we just keep bolting on pieces instead of having a modular system. What I am asking is that we have syntax for solve objects, e.g. executors

lindsayad replied:
Maybe @joshuahansel wants to work with @loganharbour to finish executors? He knocked down convergence objects

joshuahansel replied:
Feels more like they knocked _me_ down 😁  I'd be happy to try to help with executors, but Sockeye deadlines demand my attention for at least two months.

lindsayad replied:
I think in a world of executors and not this current solve object paradimg, it would be quite easy to implement multisystem. You just have a MultiSystem executor which logically would have some kind of convergence object that works naturally with multisystem. Additionally the MultiSystem or MultiXXX executor would have a parameter that takes a vector of other executors to execute. Some of these may be SingleSystem executors, for whom such a vector parameter would not make sense and for whom a multi-system or multi-executor convergence object also doesn't make sense. A MultiXXX object is fundamentally different thatn a SingleXXX object because the MultiXXX object immediately introduces the possibility of iteration. I think there should be clear distinction between objects with MultiXXX functionality and SingleXXX functionality because the parameter set is logically much more complex for the former.

GiudGiud replied:
> I think in a world of executors and not this current solve object paradigm, it would be quite easy to implement multisystem. You just have a MultiSystem executor which logically would have some kind of convergence object that works naturally with multisystem

Agree

> Additionally the MultiSystem or MultiXXX executor would have a parameter that takes a vector of other executors to execute. Some of these may be SingleSystem executors, for whom such a vector parameter would not make sense and for whom a multi-system or multi-executor convergence object also doesn't make sense

Agree

> A MultiXXX object is fundamentally different thatn a SingleXXX object because the MultiXXX object immediately introduces the possibility of iteration. I think there should be clear distinction between objects with MultiXXX functionality and SingleXXX functionality because the parameter set is logically much more complex for the former.

Agree for executors. And for solveObjects I am moving FEProblemSolve from a singleXXX to a MultiXXX, and the parameters are adapted to be vectors.

joshuahansel replied:
@lindsayad Seems like a good design at first glance - are you requesting that executors be finished before this PR? Or just discussing future refactoring?

lindsayad replied:
> are you requesting that executors be finished before this PR

I suppose not. I've been reminded of our slack history where I agreed to kick the executor can down the road. This has been a good discussion. I think I'm fine with what we're doing here. I think that I will volunteer to help @loganharbour finish executors this FY if he wants the help since I do have the bandwidth

lindsayad replied:
Before I resolve this conversation, can we do this?
```suggestion
  /// Convergence object to assess the convergence of the multi-system fixed point iteration
  const Convergence * _multi_sys_fp_convergence;
```

GiudGiud replied:
we cant, calling "isConverged" is not const

joshuahansel replied:
Just tried to remind myself why it's not const. `DefaultNonlinearConvergence` and `ReferenceResidualConvergence` are big offenders. A simple example is the ping pong iteration counter, which can get incremented in a call to `checkConvergence`.

lindsayad replied:
okie dokie

lindsayad commented on framework/src/executioners/FEProblemSolve.C:
```diff
@@ -254,60 +273,179 @@ FEProblemSolve::FEProblemSolve(Executioner & ex)
+    // being converged
+    num_fp_multisys_iters = 0;
+    converged = false;
+    while (_multi_sys_fp_convergence
+               ? (_multi_sys_fp_convergence->checkConvergence(num_fp_multisys_iters) ==
+                  Convergence::MooseConvergenceStatus::ITERATING)
+               : !converged)
```
If you don't have the convergence object and have multisystem could you iterate forever? I think that if you have multisystem, the convergence object should be a required parameter

GiudGiud replied:
See the constructor, it is required

lindsayad replied:
Maybe just add a `_have_multiple_sys` boolean then that is constructed from `_multi_sys_fp_convergence`? I think it would slightly increase readability to help dummies like me who might not read all the code better grok what's going on

lindsayad replied:
Or do something to remove this ternary? I just think this statement and the logic at the end of the loop are hard to read. It would be nice if this was just like `while (!converged)`

lindsayad replied:
Or `while (iterating)`

GiudGiud replied:
will do. Got a few commitments today so it will have to wait for tomorrow.
I m planning a follow up once https://github.com/libMesh/libmesh/pull/4007 is merged if you are ok with pushing that change until then

lindsayad replied:
sure

lindsayad reviewed (changes requested):
~I think this PR should be split into two.~ Can we change the title to highlight some of the more fundamental additions/changes in this PR? The things I'm concerned about are separate from the Physics classes.

I also do not like that we have hard-coded a certain type of fixed point iteration (for which conceptually we would hope to have something like a `FixedPointSolveObject` electable from the input file) in a `FEProblemSolve` object

GiudGiud commented:
> Can we change the title to highlight some of the more fundamental additions/changes in this PR? The things I'm concerned about are separate from the Physics classes.

Done.

> I also do not like that we have hard-coded a certain type of fixed point iteration (for which conceptually we would hope to have something like a FixedPointSolveObject electable from the input file) in a FEProblemSolve object

I have hard coded the most simple one, both conceptually and for a user (no parameters). "Electing it from the input file" is the plan for executors.
If you recall I reached out before transforming FEProblemSolve. I m happy to rework if in the end we dont like that FEProblemSolve took on a multi-system solve (even though conceptually FEProblem does not mean single-system). There are 3 options:

- create 2 new executioners, SteadyMultiSystem and TransientMultiSystem which create a different solve object. Lots of code duplication in that path imo
- what I did
- modify Steady and Transient so they do NOT always create a FEProblemSolve, but instead there's some options for the solve object, selectable from the input file. So turn this:
```
Steady::Steady(const InputParameters & parameters)
  : Executioner(parameters),
    _problem(_fe_problem),
    _feproblem_solve(*this),
    _system_time(getParam<Real>("time")),
    _time_step(_problem.timeStep()),
    _time(_problem.time()),
    _output_iteration_number(0)
{
  _fixed_point_solve->setInnerSolve(_feproblem_solve);
```
into this:
```
Steady::Steady(const InputParameters & parameters)
  : Executioner(parameters),
    _problem(_fe_problem),
    _system_time(getParam<Real>("time")),
    _time_step(_problem.timeStep()),
    _time(_problem.time()),
    _output_iteration_number(0)
{
if (_using_multi_system)
  _inner_solve = std::make_shared<MultiSystemSolve>(...);
else
  _inner_solve = std::make_shared<FEProblemSolve>(...);
```

is that preferred? That's the design I used for MultiApp fixed point iterations so I am familiar with it. 
Or do you have a 4th option?

GiudGiud committed:
Address Alex's review
- remove boolean that enables legacy behavior
- rename a parameter and re-type it to forward the selection of forward solve to the list of nonlinear systesm

Co-authored-by: Alex Lindsay <alexander.lindsay@inl.gov>

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/2554854/), step Clang format on ab87457 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/29021/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/29021/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 37b999b88168e63b60d24edc34afeda04ee57b8a`

GiudGiud commented:
@lewisgross1296

grmnptr commented on framework/src/executioners/MultiSystemSolveObject.C:
```diff
@@ -0,0 +1,60 @@
+
+MultiSystemSolveObject::MultiSystemSolveObject(Executioner & ex) : SolveObject(ex)
+{
+  // Retrieve pointers to all the systems from the problem by default
+  const auto & nl_sys_names = _problem.getNonlinearSystemNames();
+  const auto & linear_sys_names = _problem.getLinearSystemNames();
+  if (!isParamValid("system_names"))
+  {
+    for (const auto & sys_name : nl_sys_names)
+      _systems.push_back(&_problem.getSolverSystem(_problem.nlSysNum(sys_name)));
+    for (const auto & sys_name : linear_sys_names)
+      _systems.push_back(&_problem.getSolverSystem(_problem.nlSysNum(sys_name)));
+    _num_nl_systems = nl_sys_names.size();
+  }
+  else
+  {
+    _num_nl_systems = 0;
+    // Retrieve pointers to all the user-specified systems in the order that the user specified them
+    for (const auto & sys_name : getParam<std::vector<SolverSystemName>>("system_names"))
+    {
+      if (std::find(nl_sys_names.begin(), nl_sys_names.end(), sys_name) != nl_sys_names.end())
+      {
+        _systems.push_back(&_problem.getSolverSystem(_problem.nlSysNum(sys_name)));
+        _num_nl_systems++;
+      }
+      else if (std::find(linear_sys_names.begin(), linear_sys_names.end(), sys_name) !=
+               linear_sys_names.end())
+        _systems.push_back(&_problem.getSolverSystem(_problem.nlSysNum(sys_name)));
+      else
+        paramError("system_names",
+                   "System '" + sys_name +
+                       "' was not found in the Problem. Did you forget to declare it in the "
+                       "[Problem] block?");
+    }
+  }
```
```suggestion
  // Retrieve pointers to all the systems from the problem by default
  const auto & solver_sys_names = _problem.getSolverSystemNames();
  const auto & linear_sys_names = _problem.getLinearSystemNames();
  if (!isParamValid("system_names"))
  {
    Add accessor in FEProblemBase for the stored vector<SolverSystem> and just use those
  }
  else
  {
    _num_nl_systems = 0;
    // Retrieve pointers to all the user-specified systems in the order that the user specified them
    for (const auto & sys_name : getParam<std::vector<SolverSystemName>>("system_names"))
    {
      if (std::find(nl_sys_names.begin(), nl_sys_names.end(), sys_name) != nl_sys_names.end())
      {
        _systems.push_back(&_problem.getSolverSystem(_problem.solverSysNum(sys_name)));
        _num_nl_systems++;
      }
      else if (std::find(linear_sys_names.begin(), linear_sys_names.end(), sys_name) !=
               linear_sys_names.end())
        _systems.push_back(&_problem.getSolverSystem(_problem.solverSysNum(sys_name)));
      else
        paramError("system_names",
                   "System '" + sys_name +
                       "' was not found in the Problem. Did you forget to declare it in the "
                       "[Problem] block?");
    }
  }
```

grmnptr replied:
I am fine if you keep the current method to be honest I was just thinking about a loop over solver systems instead of splitting it in two. 

The more relevant parts of the change are these:
```
_systems.push_back(&_problem.getSolverSystem(_problem.solverSysNum(sys_name)));
```
because of the same numbering shift we discussed below.

GiudGiud replied:
good catch.
I think we dont need nlSysNum anymore if the "nonlinear system index" and the "solver system index" match

GiudGiud replied:
I ll need to add a test for multi-linear system to avoid these problems. I d rather do it in a later PR

Maybe a DiffusionLinearFV physics would be a good excuse to add that?

grmnptr commented on framework/src/problems/FEProblemBase.C:
```diff
@@ -6150,11 +6151,34 @@ FEProblemBase::solverSysNum(const SolverSystemName & solver_sys_name) const
+  {
+    const auto & search = _solver_sys_name_to_num.find(solver_sys_name);
+    if (search == _solver_sys_name_to_num.end())
+      mooseError("The solver system number was requested for system '" + solver_sys_name,
```
Nice!

grmnptr commented on framework/src/executioners/FEProblemSolve.C:
```diff
@@ -254,60 +268,174 @@ FEProblemSolve::FEProblemSolve(Executioner & ex)
+
+        // Call solve on the problem for that system
+        if (is_nonlinear)
+          _problem.solve(sys->number());
+        else
+          _problem.solveLinearSystem(sys->number());
```
`sys->number()` will go from 1...(N_nonlinear + N_linear) whereas both `Problem::solve()` and `Problem::solveLinearSystem` will use numbers from 1...N_nonlinear and 1...N_linear. So this would segfault I think if you use linear systems and nonlinear systems together.

lindsayad replied:
The main thrust here is correct, however at least MOOSE nonlinear system numbering (and libmesh system numbering) start at 0

grmnptr replied:
Haha, sure then the ranges are:
Nonlinear system numbering: 0...N_NLS-1 and as a solver system it is 0...N_NLS-1
Linear system numbering is: 0...N_LS-1  and as a solver system it is N_NLS...(N_NLS + N_LS - 1)
This feels a little like a bad design from my side, sorry about that.

grmnptr reviewed (changes requested):
Very nice generalization!

GiudGiud committed:
Address part of Peter's review
- fix indexing of linear systems
- doco fix

grmnptr reviewed (approved):
Looks good to me, thanks! Failure is unrelated.

I will add the example for non-physics-based multi linear problem in 
https://github.com/idaholab/moose/pull/28821
